pkg/xds/auth/k8s/authenticator.go

```diff
@@ -38,67 +40,75 @@ func (k *kubeAuthenticator) Authenticate(ctx context.Context, resource model.Res
 }
 
 func (k *kubeAuthenticator) authDataplane(ctx context.Context, dataplane *core_mesh.DataplaneResource, credential auth.Credential) error {
-	if credential == "" {
-		return errors.New("authentication failed: k8s token is missing")
+	proxyName, proxyNamespace, err := util_k8s.CoreNameToK8sName(dataplane.Meta.GetName())
+	if err != nil {
+		return err
+	}
+	serviceAccountName, err := k.podServiceAccountName(ctx, proxyName, proxyNamespace)
+	if err != nil {
+		return errors.Wrap(err, "could not retrieve Pods Service Account to verify identity of a dataplane proxy")
+	}
+	if err := k.verifyToken(ctx, credential, proxyNamespace, serviceAccountName); err != nil {
+		return errors.Wrap(err, "authentication failed")
 	}
+	return nil
+}
+
+func (k *kubeAuthenticator) verifyToken(ctx context.Context, credential auth.Credential, proxyNamespace, serviceAccountName string) error {
 	tokenReview := &kube_auth.TokenReview{
 		Spec: kube_auth.TokenReviewSpec{
 			Token: credential,
 		},
 	}
 	if err := k.client.Create(ctx, tokenReview); err != nil {
-		return errors.Wrap(err, "authentication failed: call to TokenReview API failed")
+		return errors.Wrap(err, "call to TokenReview API failed")
 	}
 	if !tokenReview.Status.Authenticated {
-		return errors.Errorf("authentication failed: token doesn't belong to a valid user")
+		return errors.Errorf("token doesn't belong to a valid user")
 	}
 	userInfo := strings.Split(tokenReview.Status.User.Username, ":")
 	if len(userInfo) != 4 {
-		return errors.Errorf("authentication failed: username inside TokenReview response has unexpected format: %q", tokenReview.Status.User.Username)
+		return errors.Errorf("username inside TokenReview response has unexpected format: %q", tokenReview.Status.User.Username)
 	}
 	if !(userInfo[0] == "system" && userInfo[1] == "serviceaccount") {
-		return errors.Errorf("authentication failed: token must belong to a k8s system account, got %q", tokenReview.Status.User.Username)
-	}
-	_, proxyNamespace, err := util_k8s.CoreNameToK8sName(dataplane.Meta.GetName())
-	if err != nil {
-		return err
+		return errors.Errorf("token must belong to a k8s system account, got %q", tokenReview.Status.User.Username)
 	}
 	namespace := userInfo[2]
 	if namespace != proxyNamespace {
-		return errors.Errorf("authentication failed: token belongs to a namespace (%q) different from proxyId (%q)", namespace, proxyNamespace)
+		return errors.Errorf("token belongs to a namespace (%q) different from proxyId (%q)", namespace, proxyNamespace)
```

Comment:
Nit: the parentheses in this error seem a bit out of place since no other error format uses them